File: urbanairship-sdk/src/main/java/com/urbanairship/iam/fullscreen/FullScreenActivity.java
/* Copyright 2018 Urban Airship and Contributors */

package com.urbanairship.iam.fullscreen;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import com.urbanairship.R;
import com.urbanairship.iam.ButtonInfo;
import com.urbanairship.iam.InAppActionUtils;
import com.urbanairship.iam.InAppMessageActivity;
import com.urbanairship.iam.ResolutionInfo;
import com.urbanairship.iam.view.InAppButtonLayout;
import com.urbanairship.iam.view.InAppViewUtils;
import com.urbanairship.iam.view.MediaView;
import com.urbanairship.widget.UAWebChromeClient;

/**
 * Full screen in-app message activity.
 */
public class FullScreenActivity extends InAppMessageActivity implements InAppButtonLayout.ButtonClickListener {

    protected FullScreenDisplayContent displayContent;
    private MediaView mediaView;

    @Override
    protected void onCreateMessage(@Nullable Bundle savedInstanceState) {
        displayContent = getMessage().getDisplayContent();
        if (displayContent == null) {
            finish();
            return;
        }

        @FullScreenDisplayContent.Template
        String template = normalizeTemplate(displayContent);

        setContentView(getTemplate(template));
        hideActionBar();

        TextView heading = findViewById(R.id.heading);
        TextView body = findViewById(R.id.body);
        InAppButtonLayout buttonLayout = findViewById(R.id.buttons);
        this.mediaView = findViewById(R.id.media);
        Button footer = findViewById(R.id.footer);
        ImageButton dismiss = findViewById(R.id.dismiss);

        // Heading
        if (displayContent.getHeading() != null) {
            InAppViewUtils.applyTextInfo(heading, displayContent.getHeading());
        } else {
            heading.setVisibility(View.GONE);
        }

        // Body
        if (displayContent.getBody() != null) {
            InAppViewUtils.applyTextInfo(body, displayContent.getBody());
        } else {
            body.setVisibility(View.GONE);
        }

        // Media
        if (displayContent.getMedia() != null) {
            mediaView.setChromeClient(new UAWebChromeClient(this));
            InAppViewUtils.loadMediaInfo(mediaView, displayContent.getMedia(), getCache());
        } else {
            mediaView.setVisibility(View.GONE);
        }

        // Button Layout
        if (!displayContent.getButtons().isEmpty()) {
            buttonLayout.setButtons(displayContent.getButtonLayout(), displayContent.getButtons());
            buttonLayout.setButtonClickListener(this);
        } else {
            buttonLayout.setVisibility(View.GONE);
        }

        // Footer
        if (displayContent.getFooter() != null) {
            InAppViewUtils.applyButtonInfo(footer, displayContent.getFooter(), 0);
            footer.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onButtonClicked(view, displayContent.getFooter());
                }
            });
        } else {
            footer.setVisibility(View.GONE);
        }

        // DismissButton
        Drawable dismissDrawable = DrawableCompat.wrap(dismiss.getDrawable()).mutate();
        DrawableCompat.setTint(dismissDrawable, displayContent.getDismissButtonColor());
        dismiss.setImageDrawable(dismissDrawable);

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getDisplayHandler().finished(ResolutionInfo.dismissed(getDisplayTime()));
                finish();
            }
        });

        // Background color
        getWindow().getDecorView().setBackgroundColor(displayContent.getBackgroundColor());
    }

    @Override
    public void onButtonClicked(View view, ButtonInfo buttonInfo) {
        InAppActionUtils.runActions(buttonInfo);
        if (buttonInfo.getBehavior().equals(ButtonInfo.BEHAVIOR_CANCEL)) {
            getDisplayHandler().cancelFutureDisplays();
        }

        getDisplayHandler().finished(ResolutionInfo.buttonPressed(buttonInfo, getDisplayTime()));
        finish();
    }

    @Override
    protected void onResume() {
        super.onResume();
        this.mediaView.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        this.mediaView.onPause();
    }

    /**
     * Gets the layout for the given template.
     *
     * @param template The fullscreen template.
     * @return The template layout resource ID.
     */
    @LayoutRes
    protected int getTemplate(@NonNull @FullScreenDisplayContent.Template String template) {
        switch (template) {
            case FullScreenDisplayContent.TEMPLATE_HEADER_BODY_MEDIA:
                return R.layout.ua_iam_fullscreen_header_body_media;

            case FullScreenDisplayContent.TEMPLATE_HEADER_MEDIA_BODY:
                return R.layout.ua_iam_fullscreen_header_media_body;

            case FullScreenDisplayContent.TEMPLATE_MEDIA_HEADER_BODY:
            default:
                return R.layout.ua_iam_fullscreen_media_header_body;
        }
    }

    /**
     * Gets the normalized template from the display content. The template may differ from the
     * display content's template to facilitate theming.
     *
     * @param displayContent The display content.
     * @return The full screen template.
     */
    @NonNull
    @FullScreenDisplayContent.Template
    protected String normalizeTemplate(FullScreenDisplayContent displayContent) {
        String template = displayContent.getTemplate();

        // If we do not have media use TEMPLATE_HEADER_BODY_MEDIA
        if (displayContent.getMedia() == null) {
            return FullScreenDisplayContent.TEMPLATE_HEADER_BODY_MEDIA;
        }

        // If we do not have a header for template TEMPLATE_HEADER_MEDIA_BODY, but we have media,
        // fallback to TEMPLATE_MEDIA_HEADER_BODY to avoid missing padding at the top modal
        if (template.equals(FullScreenDisplayContent.TEMPLATE_HEADER_MEDIA_BODY) && displayContent.getHeading() == null && displayContent.getMedia() != null) {
            return FullScreenDisplayContent.TEMPLATE_MEDIA_HEADER_BODY;
        }

        return template;
    }
}

File: urbanairship-automation/src/main/java/com/urbanairship/iam/adapter/fullscreen/FullscreenActivity.kt
/* Copyright Airship and Contributors */
package com.urbanairship.iam.adapter.fullscreen

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.view.ViewCompat
import com.urbanairship.actions.run
import com.urbanairship.automation.R
import com.urbanairship.iam.InAppMessageActivity
import com.urbanairship.iam.content.Fullscreen
import com.urbanairship.iam.content.InAppMessageDisplayContent.FullscreenContent
import com.urbanairship.iam.info.InAppMessageButtonInfo
import com.urbanairship.iam.info.InAppMessageTextInfo
import com.urbanairship.iam.view.InAppButtonLayout
import com.urbanairship.iam.view.InAppViewUtils
import com.urbanairship.iam.view.MediaView
import com.urbanairship.webkit.AirshipWebChromeClient
import kotlin.math.max

/**
 * Full screen in-app message activity.
 */
internal class FullscreenActivity : InAppMessageActivity<FullscreenContent>(), InAppButtonLayout.ButtonClickListener {

    private var mediaView: MediaView? = null
    override fun onCreateMessage(savedInstanceState: Bundle?) {
        val messageContent = this.displayContent?.fullscreen
        if (messageContent == null) {
            finish()
            return
        }

        val template = normalizeTemplate(messageContent)

        setContentView(getTemplate(template))
        hideActionBar()

        val heading: TextView = findViewById(R.id.heading)
        val body: TextView = findViewById(R.id.body)
        val buttonLayout: InAppButtonLayout = findViewById(R.id.buttons)
        val footer: Button = findViewById(R.id.footer)
        val dismiss: ImageButton = findViewById(R.id.dismiss)
        val contentHolder: View = findViewById(R.id.content_holder)
        mediaView = findViewById(R.id.media)

        // Heading
        if (messageContent.heading != null) {
            InAppViewUtils.applyTextInfo(heading, messageContent.heading)

            // Heading overlaps with the dismiss button, so it has more padding on the right than
            // the left. If it's centered, normalize the padding to prevent center text being misaligned.
            if (messageContent.heading.alignment == InAppMessageTextInfo.Alignment.CENTER) {
                normalizeHorizontalPadding(heading)
            }
        } else {
            heading.visibility = View.GONE
        }

        // Body
        if (messageContent.body != null) {
            InAppViewUtils.applyTextInfo(body, messageContent.body)
        } else {
            body.visibility = View.GONE
        }

        // Media
        mediaView?.let {
            if (messageContent.media != null) {
                it.setChromeClient(AirshipWebChromeClient(this))
                InAppViewUtils.loadMediaInfo(it, messageContent.media, assets)
            } else {
                it.visibility = View.GONE
            }
        }

        // Button Layout
        if (messageContent.buttons.isNotEmpty()) {
            buttonLayout.setButtons(messageContent.buttonLayoutType, messageContent.buttons)
            buttonLayout.setButtonClickListener(this)
        } else {
            buttonLayout.visibility = View.GONE
        }

        // Footer
        if (messageContent.footer != null) {
            InAppViewUtils.applyButtonInfo(footer, messageContent.footer, 0)
            footer.setOnClickListener { view -> onButtonClicked(view, messageContent.footer) }
        } else {
            footer.visibility = View.GONE
        }

        // Dismiss Button
        val dismissDrawable = DrawableCompat.wrap(dismiss.drawable).mutate()
        DrawableCompat.setTint(dismissDrawable, messageContent.dismissButtonColor.color)
        dismiss.setImageDrawable(dismissDrawable)
        dismiss.setOnClickListener {
            displayListener?.onUserDismissed()
            finish()
        }

        // Background color
        window.decorView.setBackgroundColor(messageContent.backgroundColor.color)

        // Apply the insets but do not consume them. Allows for the dismiss button to also receive the insets.
        if (ViewCompat.getFitsSystemWindows(contentHolder)) {
            ViewCompat.setOnApplyWindowInsetsListener(contentHolder) { v, insets ->
                ViewCompat.onApplyWindowInsets(v, insets)
                insets
            }
        }
    }

    override fun onButtonClicked(view: View, buttonInfo: InAppMessageButtonInfo) {
        buttonInfo.actions?.let {
            args.actionRunner.run(it.map)
        }
        displayListener?.onButtonDismissed(buttonInfo)
        finish()
    }

    override fun onResume() {
        super.onResume()
        mediaView?.onResume()
    }

    override fun onPause() {
        super.onPause()
        mediaView?.onPause()
    }

    /**
     * Gets the layout for the given template.
     *
     * @param template The fullscreen template.
     * @return The template layout resource ID.
     */
    @LayoutRes
    internal fun getTemplate(template: Fullscreen.Template): Int {
        return when(template) {
            Fullscreen.Template.HEADER_BODY_MEDIA -> R.layout.ua_iam_fullscreen_header_body_media
            Fullscreen.Template.HEADER_MEDIA_BODY -> R.layout.ua_iam_fullscreen_header_media_body
            Fullscreen.Template.MEDIA_HEADER_BODY -> R.layout.ua_iam_fullscreen_media_header_body
        }
    }

    /**
     * Gets the normalized template from the display content. The template may differ from the
     * display content's template to facilitate theming.
     *
     * @param fullscreen The `Fullscreen` content.
     * @return The full screen template.
     */
    private fun normalizeTemplate(fullscreen: Fullscreen): Fullscreen.Template {
        // If we do not have media use TEMPLATE_HEADER_BODY_MEDIA
        if (fullscreen.media == null) {
            return Fullscreen.Template.HEADER_BODY_MEDIA
        }

        // If we do not have a header for template TEMPLATE_HEADER_MEDIA_BODY, but we have media,
        // fallback to TEMPLATE_MEDIA_HEADER_BODY to avoid missing padding at the top modal
        return when(fullscreen.template) {
            Fullscreen.Template.HEADER_MEDIA_BODY -> {
                if (fullscreen.heading == null) {
                    Fullscreen.Template.MEDIA_HEADER_BODY
                } else {
                    fullscreen.template
                }
            }
            else -> fullscreen.template
        }
    }

    private fun normalizeHorizontalPadding(view: TextView) {
        val padding = max(ViewCompat.getPaddingEnd(view), ViewCompat.getPaddingStart(view))
        view.setPadding(padding, view.paddingTop, padding, view.paddingBottom)
        view.requestLayout()
    }
}
